// sparqlee/dist/lib/functions/OverloadTree.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OverloadTree = void 0;
const expressions_1 = require("../expressions");
const Consts_1 = require("../util/Consts");
const TypeHandling_1 = require("../util/TypeHandling");
const Helpers_1 = require("./Helpers");
/**
 * Maps argument types on their specific implementation in a tree like structure.
 * When adding any functionality to this class, make sure you add it to SpecialFunctions as well.
 */
class OverloadTree {
    constructor(identifier, depth) {
        this.identifier = identifier;
        this.implementation = undefined;
        this.subTrees = Object.create(null);
        this.depth = depth || 0;
        this.promotionCount = undefined;
    }
    /**
     * Get the implementation for the types that exactly match @param args .
     */
    getImplementationExact(args) {
        // eslint-disable-next-line @typescript-eslint/no-this-alias,consistent-this
        let node = this;
        for (const expression of args) {
            node = node.subTrees[expression];
            if (!node) {
                return undefined;
            }
        }
        return node.implementation;
    }
    getOverloadCacheIdentifier(args) {
        return this.identifier + args.map(term => {
            const literalExpression = expressions_1.isLiteralTermExpression(term);
            return literalExpression ? literalExpression.dataType : term.termType;
        }).join('');
    }
    /**
     * Searches in a depth first way for the best matching overload. considering this a the tree's root.
     * @param args:
     * @param overloadCache
     * @param superTypeProvider
     */
    search(args, superTypeProvider, overloadCache) {
        const identifier = this.getOverloadCacheIdentifier(args);
        if (overloadCache.has(identifier)) {
            return overloadCache.get(identifier);
        }
        // SearchStack is a stack of all node's that need to be checked for implementation.
        // It provides an easy way to keep order in our search.
        const searchStack = [];
        const startIndex = 0;
        if (args.length === 0) {
            return this.implementation;
        }
        // GetSubTreeWithArg return a SearchStack containing the node's that should be contacted next.
        // We also log the index since there is no other way to remember this index.
        // the provided stack should be pushed on top of our search stack since it also has it's order.
        searchStack.push(...this.getSubTreeWithArg(args[startIndex], superTypeProvider).map(node => ({ node, index: startIndex + 1 })));
        while (searchStack.length > 0) {
            const { index, node } = searchStack.pop();
            // We check the implementation because it would be possible a path is created but not implemented.
            // ex: f(double, double, double) and f(term, term). and calling f(double, double).
            if (index === args.length && node.implementation) {
                overloadCache.set(identifier, node.implementation);
                return node.implementation;
            }
            searchStack.push(...node.getSubTreeWithArg(args[index], superTypeProvider).map(item => ({ node: item, index: index + 1 })));
        }
        // Calling a function with one argument but finding no implementation should return no implementation.
        // Not even the one with no arguments.
        overloadCache.set(identifier, undefined);
        return undefined;
    }
    /**
     * Adds an overload to the tree structure considering this as the tree's root.
     * @param argumentTypes a list of ArgumentTypes that would need to be provided in the same order to
     * get the implementation.
     * @param func the implementation for this overload.
     */
    addOverload(argumentTypes, func) {
        this._addOverload([...argumentTypes], func, 0);
    }
    _addOverload(argumentTypes, func, promotionCount) {
        const [argumentType, ..._argumentTypes] = argumentTypes;
        if (!argumentType) {
            if (this.promotionCount === undefined || promotionCount <= this.promotionCount) {
                this.promotionCount = promotionCount;
                this.implementation = func;
            }
            return;
        }
        if (!this.subTrees[argumentType]) {
            this.subTrees[argumentType] = new OverloadTree(this.identifier, this.depth + 1);
        }
        this.subTrees[argumentType]._addOverload(_argumentTypes, func, promotionCount);
        // Defined by https://www.w3.org/TR/xpath-31/#promotion .
        // e.g. When a function takes a string, it can also accept a XSD_ANY_URI if it's cast first.
        // TODO: When promoting decimal type a cast needs to be preformed.
        if (argumentType === Consts_1.TypeURL.XSD_STRING) {
            this.addPromotedOverload(Consts_1.TypeURL.XSD_ANY_URI, func, arg => Helpers_1.string(arg.str()), _argumentTypes, promotionCount);
        }
        // TODO: in case of decimal a round needs to happen.
        if (argumentType === Consts_1.TypeURL.XSD_DOUBLE) {
            this.addPromotedOverload(Consts_1.TypeURL.XSD_FLOAT, func, arg => Helpers_1.double(arg.typedValue), _argumentTypes, promotionCount);
            this.addPromotedOverload(Consts_1.TypeURL.XSD_DECIMAL, func, arg => Helpers_1.double(arg.typedValue), _argumentTypes, promotionCount);
        }
        if (argumentType === Consts_1.TypeURL.XSD_FLOAT) {
            this.addPromotedOverload(Consts_1.TypeURL.XSD_DECIMAL, func, arg => Helpers_1.float(arg.typedValue), _argumentTypes, promotionCount);
        }
    }
    addPromotedOverload(typeToPromote, func, conversionFunction, argumentTypes, promotionCount) {
        if (!this.subTrees[typeToPromote]) {
            this.subTrees[typeToPromote] = new OverloadTree(this.identifier, this.depth + 1);
        }
        this.subTrees[typeToPromote]._addOverload(argumentTypes, funcConf => args => func(funcConf)([
            ...args.slice(0, this.depth),
            conversionFunction(args[this.depth]),
            ...args.slice(this.depth + 1, args.length),
        ]), promotionCount + 1);
    }
    /**
     * @param arg term to try and match to possible overloads of this node.
     * @returns SearchStack a stack with top element the next node that should be asked for implementation or overload.
     */
    getSubTreeWithArg(arg, openWorldType) {
        const res = [];
        const literalExpression = expressions_1.isLiteralTermExpression(arg);
        // These types refer to Type exported by lib/util/Consts.ts
        if (this.subTrees.term) {
            res.push(this.subTrees.term);
        }
        // TermTypes are defined in E.TermType.
        if (this.subTrees[arg.termType]) {
            res.push(this.subTrees[arg.termType]);
        }
        if (literalExpression) {
            // Defending implementation. Mainly the scary sort.
            // This function has cost O(n) + O(m * log(m)) with n = amount of overloads and m = amount of matched overloads
            // We map over each of the overloads, filter only the once that can be used (this is normally 1 or 2).
            // The sort function on an array with 1 or 2 arguments will be negligible.
            const concreteType = TypeHandling_1.isKnownLiteralType(literalExpression.dataType);
            let subExtensionTable;
            if (concreteType) {
                // Concrete dataType is known by sparqlee.
                subExtensionTable = TypeHandling_1.superTypeDictTable[concreteType];
            }
            else {
                // Datatype is a custom datatype
                subExtensionTable = TypeHandling_1.getSuperTypes(literalExpression.dataType, openWorldType);
            }
            const overLoads = Object.entries(this.subTrees);
            const matches = overLoads.filter(([matchType, _]) => matchType in subExtensionTable)
                .map(([matchType, tree]) => [subExtensionTable[matchType], tree]);
            matches.sort(([prioA, matchTypeA], [prioB, matchTypeB]) => prioA - prioB);
            res.push(...matches.map(([_, sortedType]) => sortedType));
        }
        return res;
    }
}
exports.OverloadTree = OverloadTree;
