// sparqlee/dist/lib/transformers/TermTransformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TermTransformer = void 0;
const RDFString = require("rdf-string");
const E = require("../expressions");
const Consts_1 = require("../util/Consts");
const Err = require("../util/Errors");
const P = require("../util/Parsing");
const TypeHandling_1 = require("../util/TypeHandling");
class TermTransformer {
    constructor(superTypeProvider) {
        this.superTypeProvider = superTypeProvider;
    }
    /**
     * Transforms an RDF term to the internal representation of a term,
     * assuming it is not a variable, which would be an expression (internally).
     *
     * @param term RDF term to transform into internal representation of a term
     */
    transformRDFTermUnsafe(term) {
        return this.transformTerm({
            term,
            type: 'expression',
            expressionType: 'term',
        });
    }
    transformTerm(term) {
        if (!term.term) {
            throw new Err.InvalidExpression(term);
        }
        switch (term.term.termType) {
            case 'Variable':
                return new E.Variable(RDFString.termToString(term.term));
            case 'Literal':
                return this.transformLiteral(term.term);
            case 'NamedNode':
                return new E.NamedNode(term.term.value);
            case 'BlankNode':
                return new E.BlankNode(term.term.value);
            default:
                throw new Err.InvalidTermType(term);
        }
    }
    /**
     * @param lit the rdf literal we want to transform to an internal Literal expression.
     */
    transformLiteral(lit) {
        // Both here and within the switch we transform to LangStringLiteral or StringLiteral.
        // We do this when we detect a simple literal being used.
        // Original issue regarding this behaviour: https://github.com/w3c/sparql-12/issues/112
        if (!lit.datatype || [null, undefined, ''].includes(lit.datatype.value)) {
            return lit.language ?
                new E.LangStringLiteral(lit.value, lit.language) :
                new E.StringLiteral(lit.value);
        }
        const dataType = lit.datatype.value;
        if (TypeHandling_1.isSubTypeOf(dataType, Consts_1.TypeURL.XSD_STRING, this.superTypeProvider)) {
            return new E.StringLiteral(lit.value, dataType);
        }
        if (TypeHandling_1.isSubTypeOf(dataType, Consts_1.TypeURL.RDF_LANG_STRING, this.superTypeProvider)) {
            return new E.LangStringLiteral(lit.value, lit.language);
        }
        if (TypeHandling_1.isSubTypeOf(dataType, Consts_1.TypeURL.XSD_DATE_TIME, this.superTypeProvider)) {
            // It should be noted how we don't care if its a XSD_DATE_TIME_STAMP or not.
            // This is because sparql functions don't care about the timezone.
            // It's also doesn't break the specs because we keep the string representation stored,
            // that way we can always give it back. There are also no sparql functions that alter a date.
            // (So the representation initial representation always stays valid)
            // https://github.com/comunica/sparqlee/pull/103#discussion_r688462368
            const dateVal = new Date(lit.value);
            if (Number.isNaN(dateVal.getTime())) {
                return new E.NonLexicalLiteral(undefined, dataType, this.superTypeProvider, lit.value);
            }
            return new E.DateTimeLiteral(new Date(lit.value), lit.value, dataType);
        }
        if (TypeHandling_1.isSubTypeOf(dataType, Consts_1.TypeURL.XSD_BOOLEAN, this.superTypeProvider)) {
            if (lit.value !== 'true' && lit.value !== 'false' && lit.value !== '1' && lit.value !== '0') {
                return new E.NonLexicalLiteral(undefined, dataType, this.superTypeProvider, lit.value);
            }
            return new E.BooleanLiteral(lit.value === 'true' || lit.value === '1', lit.value);
        }
        if (TypeHandling_1.isSubTypeOf(dataType, Consts_1.TypeURL.XSD_DECIMAL, this.superTypeProvider)) {
            const intVal = P.parseXSDDecimal(lit.value);
            if (intVal === undefined) {
                return new E.NonLexicalLiteral(undefined, dataType, this.superTypeProvider, lit.value);
            }
            if (TypeHandling_1.isSubTypeOf(dataType, Consts_1.TypeURL.XSD_INTEGER, this.superTypeProvider)) {
                return new E.IntegerLiteral(intVal, dataType, lit.value);
            }
            // If type is not an integer it's just a decimal.
            return new E.DecimalLiteral(intVal, dataType, lit.value);
        }
        const isFloat = TypeHandling_1.isSubTypeOf(dataType, Consts_1.TypeURL.XSD_FLOAT, this.superTypeProvider);
        const isDouble = TypeHandling_1.isSubTypeOf(dataType, Consts_1.TypeURL.XSD_DOUBLE, this.superTypeProvider);
        if (isFloat || isDouble) {
            const doubleVal = P.parseXSDFloat(lit.value);
            if (doubleVal === undefined) {
                return new E.NonLexicalLiteral(undefined, dataType, this.superTypeProvider, lit.value);
            }
            if (isFloat) {
                return new E.FloatLiteral(doubleVal, dataType, lit.value);
            }
            return new E.DoubleLiteral(doubleVal, dataType, lit.value);
        }
        return new E.Literal(lit.value, dataType, lit.value);
    }
}
exports.TermTransformer = TermTransformer;
